File: 4-sumcheck/python/prover.py
from random import randrange
from typing import Callable
from utils import arity, to_bits
from verifier import Verifier


class Prover:
    """This prover uses a function currying cache to improve its runtime"""

    def __init__(self, g, g_arity) -> None:
        """Initialize prover, compute the witness H"""
        self.g_arity = g_arity
        self.random_challenges = []
        self.cached_polynomials = [g]
        self.round = 1

        argsv = [to_bits(i, self.g_arity) for i in range(2**self.g_arity)]
        self.H = sum([self.cached_polynomials[0](*args) for args in argsv])

    def compute_and_send_next_polynomial(self, v: Verifier):
        """compute the next polynomial, append to self.polynomials, send it, update the round"""
        # make this constant, or we'll have a moving round called from within g_j
        round = self.round
        poly = self.cached_polynomials[-1]

        def g_j(X_j: int) -> int:
            pad = self.g_arity - round
            argsv = [[X_j] + to_bits(i, pad) for i in range(2**pad)]
            return sum([poly(*args) for args in argsv])
        v.receive_polynomial(g_j)
        self.round += 1

    def receive_challenge(self, challenge: int):
        """receive the latest challenge from the verifier"""
        self.random_challenges.append(challenge)
        self.cache_next(challenge)
        print("received challenge {}, initiating round {}".format(
            challenge, self.round))

    def cache_next(self, challenge: int):
        """Use the latest challenge to cache the next polynomial. Eg:
        if the last polynomial in cached polynomials is g_i(a,b,c,d)
        and the verifier just sent challenge r_i, the next cached polynomial will be
        g_{i+1}(b,c,d) = g_i(r_i,b,c,d)
        """
        poly = self.cached_polynomials[-1]

        def next_poly(*args):
            print("got args:", args)
            return poly(challenge, *args)
        self.cached_polynomials.append(next_poly)


class InefficientProver:
    def __init__(self, g, g_arity) -> None:
        """Initialize prover, compute the witness H"""
        self.g = g
        self.g_arity = g_arity
        self.random_challenges = []
        self.polynomials = []
        self.round = 1

        argsv = [to_bits(i, self.g_arity) for i in range(2**self.g_arity)]
        self.H = sum([self.g(*args) for args in argsv])

    def compute_and_send_next_polynomial(self, v: Verifier):
        """compute the next polynomial, append to self.polynomials, send it, update the round"""
        # make this constant, or we'll have a moving round called from within g_j
        round = self.round

        def g_j(X_j: int) -> int:
            args_init = self.random_challenges[:round-1] + [X_j]
            pad_len = self.g_arity - len(args_init)
            argsv = [args_init + to_bits(i, pad_len)
                     for i in range(2**pad_len)]
            # print("argsv", argsv)
            return sum([self.g(*args) for args in argsv])
        self.polynomials.append(g_j)
        v.receive_polynomial(g_j)
        self.round += 1

    def receive_challenge(self, challenge: int):
        """receive the latest challenge from the verifier"""
        self.random_challenges.append(challenge)
        print("received challenge {}, initiating round {}".format(
            challenge, self.round))

File: 4-sumcheck/src/prover.rs
use crate::{
    utils::{to_bits, FArity},
    verifier::Verifier,
    F,
};
use std::{iter::once, rc::Rc};

pub(crate) struct Prover {
    random_challenges: Vec<usize>,
    cached_polynomial: Option<FArity>,
    round: usize,
    h_claim: usize,
}

impl Prover {
    pub(crate) fn new(g: FArity ) -> Self {
        
        let h_claim = (0..2usize.pow(g.arity() as u32))
            .map(|i| g.call_f(to_bits(i, g.arity())))
            .sum();
        Self {
            random_challenges: vec![],
            cached_polynomial: Some(g),
            round: 1,
            h_claim,
        }
    }
    pub(crate) fn h_claim(&self) -> usize {
        self.h_claim
    }

    /// Need the static lifetime, so that Prover lives as long as FArity
    /// Logicking about closures is hard man
    pub(crate) fn compute_and_send_next_polynomial(&mut self, v: &mut Verifier) {
        // hand cached polynomial off to next closure
        let pad_to_len = self.cached_polynomial.as_ref().unwrap().arity() - self.round;
        let poly = self.cached_polynomial.clone().unwrap();

        let g_j: Rc<F> = Rc::new(move |X_j| {
            (1..2usize.pow(pad_to_len as u32))
                .map(|i| {
                    let args = X_j
                        .iter()
                        .cloned() // can't move out out X_j, X_j does not implement copy
                        .chain(to_bits(i, pad_to_len).into_iter())
                        .collect::<Vec<usize>>();
                    poly.call_f(args)
                })
                .sum::<usize>()
        });
        // note that g_j will always have arity 1
        let f_j = FArity::new(g_j, 1);
        v.receive_polynomial(f_j);
        self.round += 1;
        println!("PROVER: advancing to round {}", self.round);
    }
    pub(crate) fn receive_challenge(&mut self, challenge: usize) {
        self.random_challenges.push(challenge);
        self.cache_next(challenge);
    }

    fn cache_next(&mut self, challenge: usize) {
        let f = self.cached_polynomial.take().unwrap();
        let next_arity = f.arity() - 1;
        let next_f: Rc<F> = Rc::new(move |args: Vec<usize>| {
            f.call_f(once(challenge).chain(args.into_iter()).collect())
        });
        let next_farity = FArity::new(next_f, next_arity);
        self.cached_polynomial = Some(next_farity);
    }
}
